Allow subscribing with a single argument callable

composerinteralia opened the pull request:
Fixes #39976

Prior to this commit it was possible to pass a single argument block to
`ActiveSupport::Notifications.subscribe`, rather than 5 separate
arguments:

```rb
ActiveSupport::Notifications.subscribe('some_event') do |event|
  puts "Reacting to #{event.name}"
end
```

But it was not possible to do the same with a lambda, since the lambda
parameter is a required (`:req`) parameter, but we were checking only
for an optional (`:opt`) parameter.

```rb
listener = ->(event) do
  puts "Reacting to #{event.name}"
end

ActiveSupport::Notifications.subscribe('some_event', &listener)
```

It was also not possible to do this with a custom callable object, since
the custom callable does not respond directly to `:parameters` (although
it's `:call` method object does).

```rb
class CustomListener
  def call(event)
    puts "Reacting to #{event.name}"
  end
end

ActiveSupport::Notifications.subscribe('some_event', CustomListener.new)
```

Prior to this commit these examples would have raised `ArgumentError:
wrong number of arguments (given 5, expected 1)`.

With this commit the single argument lambda and custom callable work
like the single argument block.

composerinteralia commented on activesupport/lib/active_support/notifications/fanout.rb:
```diff
@@ -91,11 +91,14 @@ def self.new(pattern, listener, monotonic)
+              procish = listener.respond_to?(:parameters) ? listener : listener.method(:call)
+              params = procish.parameters
+
+              if params.length == 1 && [:opt, :req].include?(params.first.first)
```
I believe `procish.arity == 1 && procish.parameters.length == 1` would be equivalent here. Perhaps I should update to that instead?

`procish.arity` gets us most of the way there, since `proc { |*x| }` has an arity of -1 and `proc { |**x| }` has an arity of 0. The parameter length check would then cover `proc { |x, **y| }` which has an arity of 1 but a parameter length of 2.

eugeneius replied:
👍

```ruby
irb(main):001:0> proc { |x| }.then { |l| [l.arity, l.parameters.length] }
=> [1, 1]
irb(main):002:0> proc { |*x| }.then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):003:0> proc { |**x| }.then { |l| [l.arity, l.parameters.length] }
=> [0, 1]
irb(main):004:0> proc { |x, *y| }.then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):005:0> proc { |x, **y| }.then { |l| [l.arity, l.parameters.length] }
=> [1, 2]
irb(main):006:0> lambda { |x| }.then { |l| [l.arity, l.parameters.length] }
=> [1, 1]
irb(main):007:0> lambda { |*x| }.then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):008:0> lambda { |**x| }.then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):009:0> lambda { |x, *y| }.then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):010:0> lambda { |x, **y| }.then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):011:0> o = Object.new; def o.call(x); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [1, 1]
irb(main):012:0> o = Object.new; def o.call(*x); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):013:0> o = Object.new; def o.call(**x); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-1, 1]
irb(main):014:0> o = Object.new; def o.call(x, *y); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
irb(main):015:0> o = Object.new; def o.call(x, **y); end; o.method(:call).then { |l| [l.arity, l.parameters.length] }
=> [-2, 2]
```

composerinteralia replied:
Nice, thanks!

eugeneius commented on activesupport/lib/active_support/notifications/fanout.rb:
```diff
@@ -91,11 +91,14 @@ def self.new(pattern, listener, monotonic)
+            # Doing all this to detect a single argument block or callable
+            # like `proc { |x| }` vs `proc { |*x| }`, `proc { |**x| }`,
+            # or `proc { |x, **y| }`
+            if listener.respond_to?(:call)
```
`listener` has to respond to `call`, since we're always going to call `call` on it later. I don't think we need this check.

composerinteralia replied:
Yup, good ***call*** 😏

eugeneius commented on activesupport/test/notifications_test.rb:
```diff
@@ -201,6 +201,32 @@ def test_monotonic_timed_subscribed
+      ActiveSupport::Notifications.instrument("event_name")
+
+      assert_equal "event_name", event_name
+    end
```
These tests are defined in `SubscribedTest`, which is meant to test [the `subscribed` method](https://github.com/rails/rails/blob/47a4201d8dc181cab6b75cd9ffe48f339b01e3ea/activesupport/lib/active_support/notifications.rb#L242). I think they belong in `SubscribeEventObjectsTest` instead.

eugeneius reviewed (commented):
I ran into this problem today - thanks for working on this!

composerinteralia committed:
Allow subscribing with a single argument callable

Fixes #39976

Prior to this commit it was possible to pass a single argument block to
`ActiveSupport::Notifications.subscribe`, rather than 5 separate
arguments:

```rb
ActiveSupport::Notifications.subscribe('some_event') do |event|
  puts "Reacting to #{event.name}"
end
```

But it was not possible to do the same with a lambda, since the lambda
parameter is a required (`:req`) parameter, but we were checking only
for an optional (`:opt`) parameter.

```rb
listener = ->(event) do
  puts "Reacting to #{event.name}"
end

ActiveSupport::Notifications.subscribe('some_event', &listener)
```

It was also not possible to do this with a custom callable object, since
the custom callable does not respond directly to `:parameters` (although
it's `:call` method object does).

```rb
class CustomListener
  def call(event)
    puts "Reacting to #{event.name}"
  end
end

ActiveSupport::Notifications.subscribe('some_event', CustomListener.new)
```

Prior to this commit these examples would have raised `ArgumentError:
wrong number of arguments (given 5, expected 1)`.

With this commit the single argument lambda and custom callable work
like the single argument block.

composerinteralia commented:
Thanks for the review @eugeneius. Should be ready for another look.

The build failure might be a flake. I am seeing [similar errors elsewhere](https://buildkite.com/rails/rails/builds/72722#fca599ae-a0a0-4feb-a795-5980153a8c29). I can try re-running at another time, since the failing tests do involve time.

Coming back to this PR was sort of refreshing after the past 6 weeks of parental leave—it's rather rare, but my apartment is so quiet right now!

eugeneius commented:
Good thing I caught you at an opportune time 😄 Congratulations on the new pairing partner! 👶
